kimchi/src/circuits/polynomials/keccak.rs

```diff
@@ -21,65 +77,265 @@ use crate::circuits::{
+                let bnd_e = from_quarters!(bound_e, y, x);
+                let rot_e = from_quarters!(dense_rot_e, y, x);
+
+                constraints.push(
+                    word_e * T::two_pow(*off) - (quo_e.clone() * T::two_pow(64) + rem_e.clone()),
+                );
+                constraints.push(rot_e - (quo_e.clone() + rem_e));
+                constraints.push(bnd_e - (quo_e + T::two_pow(64) - T::two_pow(*off)));
+
+                for q in 0..QUARTERS {
+                    constraints.push(state_e(y, x, q) - from_shifts!(shifts_e, y, x, q));
+                    constraints.push(state_b((2 * x + 3 * y) % DIM, y, q) - expand_rot_e(y, x, q));
+                }
+            }
+        } // END pirho
+
+        // STEP chi: 4 * 5 * 5 * 3 = 300 constraints
+        for q in 0..QUARTERS {
+            for x in 0..DIM {
+                for y in 0..DIM {
+                    let not = T::literal(F::from(0x1111111111111111u64))
+                        - shifts_b(0, y, (x + 1) % DIM, q);
+                    let sum = not + shifts_b(0, y, (x + 2) % DIM, q);
+                    let and = shifts_sum(1, y, x, q);
+                    constraints.push(state_b(y, x, q) - from_shifts!(shifts_b, y, x, q));
+                    constraints.push(sum - from_shifts!(shifts_sum, y, x, q));
+                    constraints.push(state_f(y, x, q) - (shifts_b(0, y, x, q) + and));
+                }
+            }
+        } // END chi
+
+        // STEP iota: 4 constraints
+        for (q, c) in rc.iter().enumerate() {
+            constraints.push(state_g(0, 0, q) - (state_f(0, 0, q) + c.clone()));
+        } // END iota
+
+        constraints
     }
 }
 
-/// Create a Keccak rotation (whole table)
-/// Input: state (5x5) array of words to be rotated
-pub fn create_witness_keccak_rot<F: PrimeField>(state: [[u64; 5]; 5]) -> [Vec<F>; COLUMNS] {
-    // First generic gate with all zeros to constrain that the two most significant limbs of shifted output are zeros
-    let mut witness: [Vec<F>; COLUMNS] = array::from_fn(|_| vec![F::zero()]);
-    for (x, row) in ROT_TAB.iter().enumerate() {
-        for (y, &rot) in row.iter().enumerate() {
-            if rot == 0 {
-                continue;
-            }
-            rot::extend_rot(&mut witness, state[x][y], rot, RotMode::Left);
+//~
+//~ | `KeccakSponge` | [0...100) | [100...168) | [168...200) | [200...400] | [400...800) |
+//~ | -------------- | --------- | ----------- | ----------- | ----------- | ----------- |
+//~ | Curr           | old_state | new_block   | zeros       | bytes       | shifts      |
+//~ | Next           | xor_state |
+//~
+#[derive(Default)]
+pub struct KeccakSponge<F>(PhantomData<F>);
+
+impl<F> Argument<F> for KeccakSponge<F>
+where
+    F: PrimeField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::KeccakSponge);
+    const CONSTRAINTS: u32 = 568;
+
+    // Constraints for one round of the Keccak permutation function
```

Comment:
Doesn't seem right.

Reply:
Ahhh you're right, classical copy-paste typo